www/js/services.js

```diff
@@ -428,6 +429,160 @@ angular.module('emission.services', ['emission.plugin.logger'])
+      regionName: "European Union",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.14515,
+        BUS:         0.04751,
+        TRAIN:       0.048,
+        AIR_OR_HSR:  0.201
+      }
+    },
+    DE: {
+      regionName: "Germany",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.139,   // Tremod (passenger car)
+        BUS:         0.0535,  // Tremod (average city/coach)
+        TRAIN:       0.048,   // Tremod (average short/long distance)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    },
+    FR: {
+      regionName: "France",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.13125, // HBEFA (passenger car, considering 1 passenger)
+        BUS:         0.04838, // HBEFA (average short/long distance, considering 16/25 passengers)
+        TRAIN:       0.048,   // Tremod (DE value)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    },
+    AT: {
+      regionName: "Austria",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.14351, // HBEFA (passenger car, considering 1 passenger)
+        BUS:         0.04625, // HBEFA (average short/long distance, considering 16/25 passengers)
+        TRAIN:       0.048,   // Tremod (DE value)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    },
+    SE: {
+      regionName: "Sweden",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.13458, // HBEFA (passenger car, considering 1 passenger)
+        BUS:         0.04557, // HBEFA (average short/long distance, considering 16/25 passengers)
+        TRAIN:       0.048,   // Tremod (DE value)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    },
+    NO: {
+      regionName: "Norway",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.13265, // HBEFA (passenger car, considering 1 passenger)
+        BUS:         0.04185, // HBEFA (average short/long distance, considering 16/25 passengers)
+        TRAIN:       0.048,   // Tremod (DE value)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    },
+    CH: {
+      regionName: "Switzerland",
+      footprintData: {
+        WALKING:     0,
+        BICYCLING:   0,
+        CAR:         0.17638, // HBEFA (passenger car, considering 1 passenger)
+        BUS:         0.04866, // HBEFA (average short/long distance, considering 16/25 passengers)
+        TRAIN:       0.048,   // Tremod (DE value)
+        AIR_OR_HSR:  0.201    // Tremod (airplane)
+      }
+    }
+  };
+
+  var defaultCarbonDatasetCode = 'US';
+  var currentCarbonDatasetCode;
+
+  this.initialize = function() {
+    console.debug("Tiago: CarbonDatasetHelper.initialize()");
+    currentCarbonDatasetCode = defaultCarbonDatasetCode;
+    var value = KVStore.getDirect(CARBON_DATASET_KEY);
```

Comment:
Not sure `getDirect` is the right call here. We used to use local storage quite a lot, but then it turns out that iOS can delete local storage arbitrarily. So I now use `getDirect` only for state that won't annoy the user if cleared (e.g. is the walkthrough complete).

See https://github.com/e-mission/e-mission-phone/issues/443#issuecomment-423066729 and https://github.com/e-mission/e-mission-phone/pull/470 for more details on why I had to change the storage.

Reply:
Should I be using `storage.set()` and `storage.get()` instead?

Reply:
Yes.

Reply:
I can't seem to solve this in an elegant way (maybe limited knowledge of JS). Having the CarbonDatasetHelper initialize on-the-fly with a promise inside leads to the function returning with the default dataset value (US) before the promise is solved and the correct dataset code loaded from KVStore.
Would you have any suggestions on where/when to read the KVStore value beforehand? Or does it seem that I am doing something wrong, or overlooking a simple solution perhaps?

Reply:
Added a suggestion below